### PR TITLE
DragonFlyBSD support

### PR DESCRIPTION
**Pull request by tse-gratis**
Hiya, please may we upstream this DragonFlyBSD support

zig0 and zig work, and a multitude of tests pass

There are some tests that are skipped, but I would much prefer to document them as issues rather than fix them here; as that would lead to a very long delay for initial DragonFlyBSD support... and to a lot of rebasing

I hope the code is clean enough, and it can simply be merged

./zig build --build-file ../build.zig test-behavior -Dskip-release -Dskip-non-native

Gives:
> Test [166/776] behavior.byteswap.test "behavior-native-Debug-bare-multi @byteSwap vectors"...SKIP
Test [748/776] behavior.vector.test "behavior-native-Debug-bare-multi vector casts of sizes not divisable by 8"...SKIP
774 passed; 2 skipped.
Test [166/776] behavior.byteswap.test "behavior-native-Debug-c-multi @byteSwap vectors"...SKIP
Test [748/776] behavior.vector.test "behavior-native-Debug-c-multi vector casts of sizes not divisable by 8"...SKIP
774 passed; 2 skipped.
Test [166/776] behavior.byteswap.test "behavior-native-Debug-bare-single @byteSwap vectors"...SKIP
Test [748/776] behavior.vector.test "behavior-native-Debug-bare-single vector casts of sizes not divisable by 8"...SKIP
774 passed; 2 skipped.

Thankyou

### CODE REVIEW

**Comment by andrewrk** [comment]
Please see my partial review on #3559

File: src/libc_installation.hpp [hunk 1]
```diff
@@ -20,6 +20,7 @@ struct ZigLibCInstallation {
     Buf sys_include_dir;
     Buf crt_dir;
     Buf static_crt_dir;
+    Buf shared_gcc_dir;
```
Comment (andrewrk) [inline src/libc_installation.hpp]: unused field

File: lib/std/fs.zig [hunk 2]
```diff
@@ -451,6 +452,55 @@ pub const Dir = struct {
                 }
             }
 
+            fn nextDragonFly(self: *Self) !?Entry {
```
Comment (andrewrk) [inline lib/std/fs.zig]: What was different about DragonFlyBSD that nextBsd() didn't work?

File: src/link.cpp [hunk 3]
```diff
@@ -1792,28 +1792,12 @@ static void construct_linker_job_elf(LinkJob *lj) {
         if (g->libc != nullptr) {
             if (!g->have_dynamic_link) {
                 lj->args.append("--start-group");
-                if (!target_is_android(g->zig_target)) {
-                    lj->args.append("-lgcc");
-                    lj->args.append("-lgcc_eh");
-                }
```
Comment (andrewrk) [inline src/link.cpp]: Just a note to anyone else looking at this: I approved this change, it's time to lose our training wheels and rely on libcompiler_rt unconditionally. Here's to dropping support now for `-lgcc` and never looking back. :wine_glass:

File: lib/std/fs.zig (hunk 2 above)
Comment (tse-gratis) [reply lib/std/fs.zig]: DragonFlyBSD doesn't use `dirent.d_reclen`, but instead uses this macro:

> #define	_DIRENT_RECLEN(namelen) \
>	((__offsetof(struct dirent, d_name) + (namelen) + 1 + 7) & ~7)

Comment (andrewrk) [reply lib/std/fs.zig]:
```zig
pub const dirent = extern struct {
    d_fileno: c_ulong,
    d_namlen: u16,
    d_type: u8,
    d_unused1: u8,
    d_unused2: u32,
    d_name: [256]u8,

    fn reclen(self: dirent) usize {
        return (@byteOffsetOf(dirent, "d_name") + self.d_namlen + 1 + 7) & ~7;
    }
}; 
```

Then for the other BSDs where they have a `d_reclen`, add this:

```zig
    fn reclen(self: dirent) usize {
        return self.d_reclen;
    }
```

That's these files:

```
[nix-shell:~/Downloads/zig/build]$ grep -IR d_reclen ../lib/std/os/bits
../lib/std/os/bits/netbsd.zig:    d_reclen: u16,
../lib/std/os/bits/linux.zig:    d_reclen: u16,
../lib/std/os/bits/darwin.zig:    d_reclen: u16,
../lib/std/os/bits/freebsd.zig:    d_reclen: u16,
```

Comment (tse-gratis) [reply lib/std/fs.zig]: Good call. I have pushed these changes

If any other changes are needed, please go ahead and make them

File: src/link.cpp (hunk 3 above)
Comment (krytarowski) [reply src/link.cpp]: gcc is still more portable than compiler-rt here.

Comment (andrewrk) [reply src/link.cpp]: @krytarowski this is in master branch now; if you have some reasoning why `-lgcc` is ever preferable to [zig's compiler-rt](https://ziglang.org/download/0.5.0/release-notes.html#compiler-rt), I invite you to share your knowledge in a new issue

Comment (krytarowski) [reply src/link.cpp]: Do you mean that this is a builtin version from zig? If so, then this is fine!

**Pull request merged**